### PR TITLE
fby3.5: cl: Monitor PMIC error for non-shutdown case

### PR DESCRIPTION
**Pull request by DelphineChiu**
Summary:
- Monitor PMIC error after post complete.
- Add SEL to BMC when PMIC non-shutdown error happened.
- Support IPMI command to get PMIC error from BMC.

Test plan:
- Build code: Pass
- Monitor PMIC error: Pass
- Add PMIC warning SEL: Pass

Log:
1. Check BMC can get SEL when PMIC warning error happened.
(1) DIMM A0
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj VinM_OV
Error inject successfully on DIMM A0
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj high_temp
Error inject successfully on DIMM A0
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj Vin_switchover
Error inject successfully on DIMM A0
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A0
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj high_current
Error inject successfully on DIMM A0
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A0 --err_inj current_limit
Error inject successfully on DIMM A0

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (000C00) DIMM A0 High temp warning Assertion
4    slot4    2018-03-13 07:41:43    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:41:43, Sensor: PMIC_ERR (0xB4), Event Data: (000500) DIMM A0 Vin Mgmt OV Assertion
4    slot4    2018-03-13 07:41:43    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:41:43, Sensor: PMIC_ERR (0xB4), Event Data: (000B00) DIMM A0 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 07:41:43    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:41:43, Sensor: PMIC_ERR (0xB4), Event Data: (000D00) DIMM A0 Vout 1v8 PG Assertion
4    slot4    2018-03-13 07:41:43    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:41:43, Sensor: PMIC_ERR (0xB4), Event Data: (000E00) DIMM A0 High current warning Assertion
4    slot4    2018-03-13 07:41:43    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:41:43, Sensor: PMIC_ERR (0xB4), Event Data: (000F00) DIMM A0 Current limit warning Assertion

(2) DIMM A2
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj VinM_OV
Error inject successfully on DIMM A2
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj high_temp
Error inject successfully on DIMM A2
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj Vin_switchover
Error inject successfully on DIMM A2
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A2
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj high_current
Error inject successfully on DIMM A2
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A2 --err_inj current_limit
Error inject successfully on DIMM A2

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (010C00) DIMM A2 High temp warning Assertion
4    slot4    2018-03-13 08:00:48    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:00:48, Sensor: PMIC_ERR (0xB4), Event Data: (010500) DIMM A2 Vin Mgmt OV Assertion
4    slot4    2018-03-13 08:00:48    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:00:48, Sensor: PMIC_ERR (0xB4), Event Data: (010B00) DIMM A2 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 08:01:18    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:01:18, Sensor: PMIC_ERR (0xB4), Event Data: (010D00) DIMM A2 Vout 1v8 PG Assertion
4    slot4    2018-03-13 08:01:18    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:01:18, Sensor: PMIC_ERR (0xB4), Event Data: (010E00) DIMM A2 High current warning Assertion
4    slot4    2018-03-13 08:01:48    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:01:48, Sensor: PMIC_ERR (0xB4), Event Data: (010F00) DIMM A2 Current limit warning Assertion

(3) DIMM A3
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj VinM_OV
Error inject successfully on DIMM A3
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj high_temp
Error inject successfully on DIMM A3
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj Vin_switchover
Error inject successfully on DIMM A3
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A3
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj high_current
Error inject successfully on DIMM A3
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A3 --err_inj current_limit
Error inject successfully on DIMM A3

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (020C00) DIMM A3 High temp warning Assertion
4    slot4    2018-03-13 08:07:50    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:07:50, Sensor: PMIC_ERR (0xB4), Event Data: (020500) DIMM A3 Vin Mgmt OV Assertion
4    slot4    2018-03-13 08:07:50    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:07:50, Sensor: PMIC_ERR (0xB4), Event Data: (020B00) DIMM A3 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 08:07:50    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:07:50, Sensor: PMIC_ERR (0xB4), Event Data: (020D00) DIMM A3 Vout 1v8 PG Assertion
4    slot4    2018-03-13 08:07:50    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:07:50, Sensor: PMIC_ERR (0xB4), Event Data: (020E00) DIMM A3 High current warning Assertion
4    slot4    2018-03-13 08:13:52    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:13:52, Sensor: PMIC_ERR (0xB4), Event Data: (020F00) DIMM A3 Current limit warning Assertion

(4) DIMM A4
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj VinM_OV
Error inject successfully on DIMM A4
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj high_temp
Error inject successfully on DIMM A4
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj Vin_switchover
Error inject successfully on DIMM A4
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A4
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj high_current
Error inject successfully on DIMM A4
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A4 --err_inj current_limit
Error inject successfully on DIMM A4

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (030C00) DIMM A4 High temp warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030500) DIMM A4 Vin Mgmt OV Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030B00) DIMM A4 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030D00) DIMM A4 Vout 1v8 PG Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030E00) DIMM A4 High current warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030F00) DIMM A4 Current limit warning Assertion

(5) DIMM A6
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj VinM_OV
Error inject successfully on DIMM A6
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj high_temp
Error inject successfully on DIMM A6
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj Vin_switchover
Error inject successfully on DIMM A6
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A6
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj high_current
Error inject successfully on DIMM A6
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A6 --err_inj current_limit
Error inject successfully on DIMM A6

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (040C00) DIMM A6 High temp warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030500) DIMM A4 Vin Mgmt OV Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030B00) DIMM A4 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030D00) DIMM A4 Vout 1v8 PG Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030E00) DIMM A4 High current warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (030F00) DIMM A4 Current limit warning Assertion

(6) DIMM A7
- Error inject
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj VinM_OV
Error inject successfully on DIMM A7
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj high_temp
Error inject successfully on DIMM A7
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj Vin_switchover
Error inject successfully on DIMM A7
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj Vout_1v8_PG
Error inject successfully on DIMM A7
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj high_current
Error inject successfully on DIMM A7
root@bmc-oob:~# dimm-util slot4 --pmic --dimm A7 --err_inj current_limit
Error inject successfully on DIMM A7

- Check SEL
root@bmc-oob:~# log-util --print slot4 | grep SEL
...
4    slot4    2018-03-13 07:37:12    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 07:37:12, Sensor: PMIC_ERR (0xB4), Event Data: (050C00) DIMM A7 High temp warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (050500) DIMM A7 Vin Mgmt OV Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (050B00) DIMM A7 Vin Mgmt to Vin buck switchover Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (050D00) DIMM A7 Vout 1v8 PG Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (050E00) DIMM A7 High current warning Assertion
4    slot4    2018-03-13 08:11:21    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:11:21, Sensor: PMIC_ERR (0xB4), Event Data: (050F00) DIMM A7 Current limit warning Assertion

2. Check BIC can still monitor PMIC error after host cycle.
4    slot4    2018-03-13 08:18:14    gpiod            FRU: 4, System powered OFF
4    slot4    2018-03-13 08:18:19    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:18:19, Sensor: ME_POWER_STATE (0x16), Event Data: (000000) RUNNING Assertion
4    slot4    2018-03-13 08:18:20    power-util       SERVER_POWER_CYCLE successful for FRU: 4
4    slot4    2018-03-13 08:18:24    gpiod            FRU: 4, System powered ON
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (000F00) DIMM A0 Current limit warning Assertion
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (010F00) DIMM A2 Current limit warning Assertion
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (020F00) DIMM A3 Current limit warning Assertion
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (030F00) DIMM A4 Current limit warning Assertion
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (040F00) DIMM A6 Current limit warning Assertion
4    slot4    2018-03-13 08:19:30    ipmid            SEL Entry: FRU: 4, Record: Standard (0x02), Time: 2018-03-13 08:19:30, Sensor: PMIC_ERR (0xB4), Event Data: (050F00) DIMM A7 Current limit warning Assertion
4    slot4    2018-03-13 08:19:35    ipmid            SEL Entry: FRU: 4, Record: Facebook Unified SEL (0xFB), GeneralInfo: POST(0x28), POST Failure Event: System PXE boot fail
4    slot4    2018-03-13 08:20:36    ipmid            SEL Entry: FRU: 4, Record: Facebook Unified SEL (0xFB), GeneralInfo: POST(0x28), POST Failure Event: System PXE boot fail

### CODE REVIEW

File: meta-facebook/yv35-cl/src/platform/plat_pmic.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+int get_dimm_info(uint8_t dimm_id, uint8_t *bus, uint8_t *addr)
+{
+	int ret = 0;
+
+	if ((bus == NULL) || (addr == NULL)) {
+		return -1;
+	}
+
+	if ((dimm_id < 0) || (dimm_id >= MAX_COUNT_DIMM)) {
```
Comment (garnermic) [inline meta-facebook/yv35-cl/src/platform/plat_pmic.c:124]: dimm_id is a uint8_t so can never be <0.  Please fix this test to remove the <0 test.  Thanks!

**Review by garnermic — CHANGES_REQUESTED** [review]
One minor change requested upon internal review.

**Comment by SaraLin-wiwynn** [comment]
Hi @garnermic ,
Thanks for your comments.
Based on your suggestions to modify the code, rebase main and force update this branch.
Please help review it again, thank you.

**Comment by SaraLin-wiwynn** [comment]
Hi @garnermic ,
We fix the problem about repeating add SEL when DIMM is not present and force update this branch.
Please help review it again, thank you.

**Comment by LoraLin1** [comment]
Hi @garnermic,
We have fixed the conflict and do the force update.
Please help to review again, thank you.

**Pull request closed without merging**